### PR TITLE
Recommendation algo

### PR DESCRIPTION
**Pull request by Fnuworsu**
# Description  
- What does this PR do?  
  `Implements a custom Priority Queue data structure for the recommendation algorithm`  
- Why is it necessary or valuable?  
  `Provides an efficient O(log n) implementation for sorting and retrieving top video recommendations`  
- What is intentionally left out and will be done in a later PR?  
  `Performance optimizations for large datasets and additional heap operations`

---

# Milestones / Related Work  
- Feature or user story this contributes to: `Technical Challenge #1 Post recommendation algorithm based on youtube content`  
- Milestone or version target: `Base implementation that recommends topK (K== 10 by default) posts`  
---

# Resources and References  
- Tutorials, documentation, or code examples used:  
  ``Binary Heap implementation concepts from "Introduction to Algorithms``
[Geeks for geeks article I read](https://www.geeksforgeeks.org/dsa/binary-heap/)  
- Any copied or adapted code (with attribution):  
  `None, custom implementation`  
---

# Test Plan  
- How was this tested?  
  `Manual tests and integration with sample video data through test.js`  
- Screenshots or recordings:  
Screenshot 2025-07-10 at 1 42 37 PM

- Seed or mock data used:  
  `100 sample video entries in sample.js`

## Edge Cases Tested  
*   Empty priority queue
*   Single element queue
*   Duplicate priority values
*   Custom comparison function`

---

# Additional Notes  
- Known limitations: `Currently optimized for in-memory operations, will need persistence for larger datasets`   
- Reviewer requests or feedback needed: `Feedback on the comparison function implementation and overall heap efficiency`

### CODE REVIEW

File: backend/recommendationAlgo/priorityQueue.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        this.compare = compareFunction ?? ((a,b) => (a > b))
+        this.priorityQueue = []
+    }
+
+    size() {
+        return this.priorityQueue.length
+    }
+
+    isEmpty() {
+        return this.priorityQueue.length === 0
```
Comment (dorcheng) [inline backend/recommendationAlgo/priorityQueue.js:18]: for this, you can leverage the size() method you implemented and do

```
return this.size() === 0
```

File: backend/recommendationAlgo/priorityQueue.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            const p = this.#parent(i)
+            if (this.compare(this.priorityQueue[p], this.priorityQueue[i])) break
+            this.#swap(i,p)
+            i = p
+        }
+    }
+
+    #bubbleDown(i) {
+        const len = this.priorityQueue.length
+        while (true) {
```
Comment (dorcheng) [inline backend/recommendationAlgo/priorityQueue.js:66]: hmm is there a way for you to not use `while(true)` and instead specify a specific condition inside the while loop? Using `while(true)` is generally not the best practice because you can easily make a mistake and accidentally cause the code to enter an infinite loop!

File: backend/recommendationAlgo/priorityQueue.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        this.priorityQueue.push(value)
+        this.#bubbleUp(this.priorityQueue.length - 1)
+    }
+    pop() {
+        const len = this.priorityQueue.length
+        // base cases
+        if (len === 0) return null
+        if (len === 1) return this.priorityQueue.pop()
+
+        const top = this.priorityQueue[0]
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/recommendationAlgo/priorityQueue.js:33]: True, doing this will make my code less repetitive. Thanks for the heads up
